### DISCUSSION TITLE
Improving support for JSON files as data sources for BIRT

### DISCUSSION THREAD

**Discussion by Markus-Elfring** [eclipse-birt/birt#985] [Q&A]
The “Business Intelligence Reporting Tool 4.9” supports also the following data sources.
* XML
* Scripted variant

:crystal_ball: Will the support evolve any further for the data format “JavaScript Object Notation” in this context?

Would you like to share experiences for corresponding data conversions? :thinking:

**Comment by wimjongman** [comment]
> The “Business Intelligence Reporting Tool 4.9” supports also the following data sources.
> 
>     * XML
> 
>     * Scripted variant
> 
> 
> 🔮 Will the support evolve any further for the data format “JavaScript Object Notation” in this context?
> 
> Would you like to share experiences for corresponding data conversions? 🤔

We are probably not going to make this. But if you want to contribute, then please give it a shot.

Maybe you just want to convert the JSON to XML:

```
Use the (excellent) JSON-Java library from json.org then

JSONObject json = new JSONObject(str);
String xml = XML.toString(json);
```

Or even wrap it in a POJO datasource.

**Comment by claesrosell** [comment]
I will probably develop JSON ODA support for Birt since my company will need that in the future. I have not decided on a query format yet though.

**Comment by TheLaserNinja** [comment]
I have a JSON emitter. I have written a JSON ODA in the past, however JSON is best done through a scripted data object. It's just a lot more flexible. Just use a Javascript EVAL statement in your scripted data object, and you're all set. We may have included the Javascript parse json function in newer versions of the BIRT script engine, however, outside of the other JSON methods in the library, it doesnt do anything extra beyond what an eval statement will.

**Reply by hvbtup** [reply]
Isn't using eval for parsing JSON considered bad practice regarding security?

**Reply by TheLaserNinja** [reply]
Depends on several things. There are risks that would allow for things like infinite loops for crashing it or running malicious code if a bad actor could feed to it. This is why you never use eval in any user input forms. However, if you know your data source is clean or validate the JSON beforehand, then there is no real risk. 

If something like JSON.parse() is available, it is preferable as it will only process valid JSON. But if you don't have JSON.parse() available eval will do in a pinch.
